File: msdialog/src/main/java/com/maple/msdialog/AlertNumberPickerDialog.java
package com.maple.msdialog;

import android.app.Dialog;
import android.content.Context;
import android.graphics.Typeface;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.DatePicker;
import android.widget.ImageView;
import android.widget.NumberPicker;
import android.widget.TextView;


/**
 * 警告框式数字选择器Dialog [ 标题 + 数字选择 + 左按钮 + 右按钮 ]
 *
 * @author maple
 * @time 2018/12/6
 */
public class AlertNumberPickerDialog extends BaseDialog {
    private TextView txt_title;
    private NumberPicker np_number;
    private TextView tv_suffix;
    private TextView leftButton;
    private TextView rightButton;
    private ImageView img_line;

    private boolean showTitle = false;
    private boolean showMsg = false;
    private boolean showRightBtn = false;
    private boolean showLeftBtn = false;


    public AlertNumberPickerDialog(Context context) {
        super(context);
        rootView = LayoutInflater.from(context).inflate(R.layout.view_number_picker_dialog, null);

        // get custom Dialog layout
        txt_title = rootView.findViewById(R.id.txt_title);
        txt_title.setVisibility(View.GONE);
        np_number = rootView.findViewById(R.id.np_number);
        tv_suffix = rootView.findViewById(R.id.tv_suffix);
        tv_suffix.setVisibility(View.GONE);
        leftButton = rootView.findViewById(R.id.tv_left);
        leftButton.setVisibility(View.GONE);
        rightButton = rootView.findViewById(R.id.tv_right);
        rightButton.setVisibility(View.GONE);
        img_line = rootView.findViewById(R.id.img_line);
        img_line.setVisibility(View.GONE);

        // set Dialog style
        dialog = new Dialog(context, R.style.AlertDialogStyle);
        dialog.setContentView(rootView);

        setScaleWidth(0.8);
    }

    public AlertNumberPickerDialog setCancelable(boolean cancel) {
        dialog.setCancelable(cancel);
        return this;
    }

    @Override
    public AlertNumberPickerDialog setScaleWidth(double scWidth) {
        return (AlertNumberPickerDialog) super.setScaleWidth(scWidth);
    }

    public AlertNumberPickerDialog setTitle(String title) {
        int color = ContextCompat.getColor(mContext, R.color.def_title_color);
        return setTitle(title, color, 18, false);
    }

    public AlertNumberPickerDialog setTitle(String title, int color, int size, boolean isBold) {
        showTitle = true;
        txt_title.setText(title);
        if (color != -1) {
            txt_title.setTextColor(color);
        }
        if (size > 0) {
            txt_title.setTextSize(size);
        }
        if (isBold) {
            txt_title.setTypeface(txt_title.getTypeface(), Typeface.BOLD);
        }
        return this;
    }

    public AlertNumberPickerDialog setNumberValues(String[] displayedValues, int defaultValue, NumberPicker.OnValueChangeListener onValueChangedListener) {
        showMsg = true;
        np_number.setDisplayedValues(displayedValues);
        np_number.setMinValue(0);
        np_number.setMaxValue(displayedValues.length - 1);
        // 设置为对当前值不可编辑
        np_number.setDescendantFocusability(DatePicker.FOCUS_BLOCK_DESCENDANTS);
        // default value
        np_number.setValue(defaultValue);
        np_number.setOnValueChangedListener(onValueChangedListener);
        return this;
    }

    public AlertNumberPickerDialog setNumberValueSuffix(String suffix) {
        int color = ContextCompat.getColor(mContext, R.color.def_title_color);
        return setNumberValueSuffix(suffix, color, 16);
    }

    public AlertNumberPickerDialog setNumberValueSuffix(String suffix, int color, int size) {
        showMsg = true;
        tv_suffix.setText(suffix);
        if (color != -1) {
            tv_suffix.setTextColor(color);
        }
        if (size > 0) {
            tv_suffix.setTextSize(size);
        }
        tv_suffix.setVisibility(View.VISIBLE);
        return this;
    }

    public AlertNumberPickerDialog setRightButton(String text, final OnClickListener listener) {
        int color = ContextCompat.getColor(mContext, R.color.def_title_color);
        return setRightButton(text, color, 16, false, listener);
    }

    public AlertNumberPickerDialog setRightButton(String text, int color, int size, boolean isBold, final OnClickListener listener) {
        showRightBtn = true;
        if (TextUtils.isEmpty(text)) {
            rightButton.setText(R.string.ok);
        } else {
            rightButton.setText(text);
        }
        if (color != -1) {
            rightButton.setTextColor(color);
        }
        if (size > 0) {
            rightButton.setTextSize(size);
        }
        if (isBold) {
            rightButton.setTypeface(rightButton.getTypeface(), Typeface.BOLD);
        }
        rightButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                if (listener != null) {
                    listener.onClick(v);
                }
                dialog.dismiss();
            }
        });
        return this;
    }

    public AlertNumberPickerDialog setLeftButton(String text, final OnClickListener listener) {
        int color = ContextCompat.getColor(mContext, R.color.def_title_color);
        return setLeftButton(text, color, 16, false, listener);
    }

    public AlertNumberPickerDialog setLeftButton(String text, int color, int size, boolean isBold, final OnClickListener listener) {
        showLeftBtn = true;
        if (TextUtils.isEmpty(text)) {
            leftButton.setText(R.string.cancel);
        } else {
            leftButton.setText(text);
        }
        if (color != -1) {
            leftButton.setTextColor(color);
        }
        if (size > 0) {
            leftButton.setTextSize(size);
        }
        if (isBold) {
            leftButton.setTypeface(leftButton.getTypeface(), Typeface.BOLD);
        }
        leftButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                if (listener != null) {
                    listener.onClick(v);
                }
                dialog.dismiss();
            }
        });
        return this;
    }

    private void setLayout() {
        if (!showTitle && !showMsg) {
            txt_title.setText(R.string.alert);
            txt_title.setVisibility(View.VISIBLE);
        }
        if (showTitle) {
            txt_title.setVisibility(View.VISIBLE);
        }
        // one button
        if (!showRightBtn && !showLeftBtn) {
            rightButton.setText(R.string.ok);
            rightButton.setVisibility(View.VISIBLE);
            rightButton.setBackgroundResource(R.drawable.alertdialog_single_selector);
            rightButton.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    dialog.dismiss();
                }
            });
        }

        if (showRightBtn && !showLeftBtn) {
            rightButton.setVisibility(View.VISIBLE);
            rightButton.setBackgroundResource(R.drawable.alertdialog_single_selector);
        }

        if (!showRightBtn && showLeftBtn) {
            leftButton.setVisibility(View.VISIBLE);
            leftButton.setBackgroundResource(R.drawable.alertdialog_single_selector);
        }
        // two button
        if (showRightBtn && showLeftBtn) {
            rightButton.setVisibility(View.VISIBLE);
            rightButton.setBackgroundResource(R.drawable.alertdialog_right_selector);
            leftButton.setVisibility(View.VISIBLE);
            leftButton.setBackgroundResource(R.drawable.alertdialog_left_selector);
            img_line.setVisibility(View.VISIBLE);
        }
    }

    public void show() {
        setLayout();
        dialog.show();
    }

}

File: msdialog/src/main/java/com/maple/msdialog/AlertNumberPickerDialog.kt
package com.maple.msdialog

import android.content.Context
import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.widget.DatePicker
import android.widget.NumberPicker.OnValueChangeListener
import androidx.annotation.ColorInt
import androidx.annotation.StyleRes
import androidx.core.content.ContextCompat
import com.maple.msdialog.databinding.MsDialogNumberPickerBinding
import com.maple.msdialog.utils.DensityUtils.dp2px

/**
 * 警告框式数字选择器Dialog [ 标题 + 数字选择 + 左按钮 + 右按钮 ]
 *
 * @author shaoshuai
 * @time 2018/12/6
 */
class AlertNumberPickerDialog(
    private val mContext: Context,
    private val config: Config = Config(mContext),
    @StyleRes themeResId: Int = R.style.AlertDialogStyle
) : AlertBaseDialog(mContext, config, themeResId) {
    private val binding: MsDialogNumberPickerBinding = MsDialogNumberPickerBinding.inflate(
        LayoutInflater.from(mContext), null, false
    )
    private var showTitle = false
    private var showMsg = false

    // 方便java调用
    constructor(mContext: Context) : this(mContext, Config(mContext), R.style.AlertDialogStyle)
    constructor(mContext: Context, config: Config) : this(mContext, config, R.style.AlertDialogStyle)

    fun getTitleView() = binding.tvTitle
    fun getSuffixView() = binding.tvSuffix

    fun setDialogTitle(title: CharSequence?): AlertNumberPickerDialog {
        return setTitle(title, isBold = false)
    }

    override fun setTitle(title: CharSequence?) {
        this.setTitle(title, isBold = false)
    }

    fun setTitle(
        title: CharSequence?,
        @ColorInt color: Int = config.titleColor,// 字体颜色
        spSize: Float = config.titleTextSizeSp,// 字体大小
        isBold: Boolean = false
    ): AlertNumberPickerDialog {
        showTitle = true
        getTitleView().apply {
            text = title ?: config.defNullText
            setTextColor(color)
            textSize = spSize
            setPadding(config.titlePaddingLeft, config.titlePaddingTop, config.titlePaddingRight, config.titlePaddingBottom)
            setTypeface(typeface, if (isBold) Typeface.BOLD else Typeface.NORMAL)
        }
        return this
    }

    fun setNumberValues(displayedValues: Array<String>, defaultValue: Int, onValueChangedListener: OnValueChangeListener?): AlertNumberPickerDialog {
        showMsg = true
        binding.npNumber.apply {
            this.displayedValues = displayedValues
            minValue = 0
            maxValue = displayedValues.size - 1
            // 设置为对当前值不可编辑
            descendantFocusability = DatePicker.FOCUS_BLOCK_DESCENDANTS
            // default value
            value = defaultValue
            setOnValueChangedListener(onValueChangedListener)
        }
        return this
    }

    // 设置value的后缀
    fun setNumberValueSuffix(
        suffix: CharSequence?,
        @ColorInt color: Int = config.messageColor,// 字体颜色
        spSize: Float = config.messageTextSizeSp // 字体大小
    ): AlertNumberPickerDialog {
        showMsg = true
        getSuffixView().apply {
            text = suffix ?: config.defNullText
            setTextColor(color)
            textSize = spSize
            visibility = View.VISIBLE
        }
        return this
    }

    override fun setLayout() {
        super.setLayout()
        // title & msg
        setDialogContext(binding.root)
        getTitleView().visibility = if (showTitle) View.VISIBLE else View.GONE
        if (!showTitle && !showMsg) {
            getTitleView().text = config.defNullText
            getTitleView().visibility = View.VISIBLE
        }
    }

    /**
     * AlertNumberPickerDialog 的配置
     */
    open class Config(
        context: Context
    ) : AlertBaseDialog.Config(context) {
        // title
        var titleTextSizeSp: Float = 18f // 字体大小
        @ColorInt var titleColor: Int = ContextCompat.getColor(context, R.color.ms_def_title_color) // 字体颜色
        var titlePaddingLeft: Int = 15f.dp2px(context) // 左边距
        var titlePaddingTop: Int = 22f.dp2px(context) // 上边距
        var titlePaddingRight: Int = 15f.dp2px(context) // 右边距
        var titlePaddingBottom: Int = 0f.dp2px(context) // 下边距

        // message
        var messageTextSizeSp: Float = 16f // 字体大小
        @ColorInt var messageColor: Int = ContextCompat.getColor(context, R.color.ms_def_message_color) // 字体颜色
    }

}